feat(gnoweb): mention markdown ext

alexiscolin opened the pull request:
This PR adds support for detecting and converting `@username `mentions and `gno addresses` into clickable links in markdown content. 

## Features

### `@ Mentions`

- Detects @username patterns where username contains alphanumeric characters and underscores
- Validates word boundaries to avoid matching email addresses
- Converts valid mentions to /u/username links
- Supports usernames with underscores (e.g., @user_name_123)
- Prevents matching in contexts like user@domain.com (emails) or embedded addresses

### `g1 Addresses`

- Detects and validates bech32 addresses starting with g1
- Converts valid addresses to /u/address links

## Implementation
### Architecture

- Inline Parser: Handles `@` mentions with trigger-based detection
- Inline Parser: Handle ` ` (space) followed by a g1 in a regex ([since there is no way to trigger  alphanumerics chars are with Goldmark](https://pkg.go.dev/github.com/yuin/goldmark/parser#InlineParser))
- Shared Logic: Common link creation and boundary validation

### Link Types
- Added new user link type with icon (WIP - to define) for mentions and addresses (in addition to inter-realm / external / action links)
- Links point to /u/ routes for user profiles

Gno2D2 commented:
#### 🛠 PR Checks Summary
 🔴 Changes related to gnoweb must be reviewed by its codeowners
 🔴 Pending initial approval by a review team member, or review from tech-staff


##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🔴 Changes related to gnoweb must be reviewed by its codeowners
 🔴 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: alexiscolin/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Changes related to gnoweb must be reviewed by its codeowners</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 A changed file matches this pattern: ^gno.land/pkg/gnoweb/ (filename: gno.land/pkg/gnoweb/markdown/ext.go)

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 Or
    ├── 🔴 Or
    │   ├── 🔴 And
    │   │   ├── 🟢 Pull request author is user: alexiscolin
    │   │   └── 🔴 This user reviewed pull request: gfanton (with state "APPROVED")
    │   └── 🔴 And
    │       ├── 🔴 Pull request author is user: gfanton
    │       └── 🔴 This user reviewed pull request: alexiscolin (with state "APPROVED")
    └── 🔴 And
        ├── 🔴 Not (🟢 Pull request author is user: alexiscolin)
        ├── 🟢 Not (🔴 Pull request author is user: gfanton)
        └── 🔴 Or
            ├── 🔴 This user reviewed pull request: alexiscolin (with state "APPROVED")
            └── 🔴 This user reviewed pull request: gfanton (with state "APPROVED")

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🔴 Requirement not satisfied
└── 🔴 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🔴 At least 1 user(s) of the organization reviewed the pull request (with state "APPROVED")
    │       ├── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🟢 This pull request is a draft
    └── 🔴 Then
        └── 🔴 And
            ├── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)
            └── 🔴 At least 1 user(s) of the team tech-staff reviewed pull request

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

alexiscolin commented:
I cannot use the inline parser to trigger on g directly, as ([there is no way to trigger alphanumerics chars are with Goldmark](https://pkg.go.dev/github.com/yuin/goldmark/parser#InlineParser)). Instead, I replaced the AST transformer with an inline parser that uses a space as the trigger and checks for a g1 prefix in the condition. This approach works and should be more efficient.
I’ve updated the PR description accordingly.

gfanton commented:
I see, but using spaces will unfortunately prevent people from using an address at the beginning of the line. So I'm afraid we need to fall back to your original solution. What I suggest is keeping inline parsing for `@` mentions and using a transformer for `g1...` address targeting specific text blocks such as paragraphs, headings, etc...

alexiscolin commented:
> I see, but using spaces will unfortunately prevent people from using an address at the beginning of the line. So I'm afraid we need to fall back to your original solution. What I suggest is keeping inline parsing for `@` mentions and using a transformer for `g1...` address targeting specific text blocks such as paragraphs, headings, etc...

Should work like that. `g` detected in addition to ` ` but only if at beginning of the line.

leohhhn commented:
I don't think we should try linking things that are already linked, as so:

<img width="613" height="144" alt="Screenshot 2025-07-21 at 00 40 00" src="https://github.com/user-attachments/assets/0ec61407-e17d-4645-8ef3-fc92cc54bb4a" />

gfanton reviewed (approved):
working great !
